### ISSUE TITLE
Installing Tokenizers without Internet connection

### ISSUE THREAD

**Issue by jordiae** [huggingface/tokenizers#176]
Hi, I'm trying to install the Tokenizers package in a cluster that doesn't have access to the Internet. Specifically, I'm interested in the Python bindings. So far, I have installed some dependencies, but when I run the setup.py script I get the following log:

```
running install
running bdist_egg
running egg_info
writing tokenizers.egg-info/PKG-INFO
writing dependency_links to tokenizers.egg-info/dependency_links.txt
writing top-level names to tokenizers.egg-info/top_level.txt
reading manifest file 'tokenizers.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '*' found under directory 'tokenizers-lib/target'
writing manifest file 'tokenizers.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-ppc64le/egg
running install_lib
running build_py
running build_ext
running build_rust
    Updating crates.io index
```

It gets stuck since it has no connection but it's still trying to update the crates.io index.

Any suggestions on how could I prevent the script from trying to connect to the internet or which dependencies should I install or configure locally?

Unlike a [recent, related issue](https://github.com/huggingface/tokenizers/issues/173), in my case, it's a Power9 architecture so there is no wheel available and I must install directly from the tar file. 

TL;DR: How to install Tokenizers if `pip install tokenizers` is not an option because there is no connection to the Internet.

Many thanks in advance.

**Comment by n1t0** [comment]
Hi @jordiae!

I think you should be able to use the `cargo vendor` command to actually include all the dependencies in the repository before you move it. Can you try doing the following:
1. Extract `tokenizers` somewhere on a computer with internet access
2. `cd bindings/python`
3. `mkdir .cargo`
4. `cargo vendor > .cargo/config`

You can then zip the entire repository again and transfer it to your server. You will also need to install the python package `setuptools_rust` on your server in order to be able to compile the Python bindings. For this one you should be able to use the [provided wheel](https://pypi.org/project/setuptools-rust/#files) since it is just Python code. Then on your server:
1. Extract your zip wherever relevant
2. `cd bindings/python`
3. `cargo build --offline`
4. `python setup.py install`

If you are using a Python virtual env, you just need to make sure you activated it before running the last command. Let me know how this goes!

**Comment by jordiae** [comment]
Hi @n1t0! Many thanks for your answer. On my server, when executing `cargo build --offline`, I get the following error:
`error: failed to run custom build command for "pyo3 v0.8.4"`
`thread "main" panicked at 'Error: pyo3 requires a nightly or dev version of Rust.`

**Comment by n1t0** [comment]
Indeed, you need to have a nightly version of Rust installed for the Python bindings. You should be able to find a standalone installer for the nightly version, for your current platform here: https://forge.rust-lang.org/infra/other-installation-methods.html

**Comment by jordiae** [comment]
@n1t0 Perfect, already working on it. Many thanks for your answers!

**Comment by ppiyush28** [comment]
Hi @n1t0 

After python setup.py install, I got the below error : 

Please help to run tokenizer & transformer.

error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
   --> /Users/piyushpande/tokenizers/tokenizers/src/models/bpe/trainer.rs:541:47
    |
537 |                     let w = &words[*i] as *const _ as *mut _;
    |                             -------------------------------- casting happend here
...
541 |                         let word: &mut Word = &mut (*w);
    |                                               ^^^^^^^^^
    |
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
    = note: `#[deny(invalid_reference_casting)]` on by default

error: could not compile `tokenizers` (lib) due to 1 previous error

Caused by:
  process didn't exit successfully: `/Users/piyushpande/.rustup/toolchains/stable-aarch64-apple-darwin/bin/rustc --crate-name tokenizers --edition=2018 /Users/piyushpande/tokenizers/tokenizers/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=133 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="cached-path"' --cfg 'feature="clap"' --cfg 'feature="cli"' --cfg 'feature="default"' --cfg 'feature="dirs"' --cfg 'feature="esaxx_fast"' --cfg 'feature="http"' --cfg 'feature="indicatif"' --cfg 'feature="onig"' --cfg 'feature="progressbar"' --cfg 'feature="reqwest"' -C metadata=ae5df84b17f0adbf -C extra-filename=-ae5df84b17f0adbf --out-dir /Users/piyushpande/tokenizers/bindings/python/target/release/deps -C strip=debuginfo -L dependency=/Users/piyushpande/tokenizers/bindings/python/target/release/deps --extern aho_corasick=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libaho_corasick-59b5ae5ee817a67b.rmeta --extern cached_path=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libcached_path-659cef5bd8f42526.rmeta --extern clap=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libclap-e822628a367d4ba0.rmeta --extern derive_builder=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libderive_builder-a0421aaff6307ba8.rmeta --extern dirs=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libdirs-c5979ca97e895266.rmeta --extern esaxx_rs=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libesaxx_rs-775e025381693d67.rmeta --extern getrandom=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libgetrandom-f931a4ae6cb1a28a.rmeta --extern indicatif=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libindicatif-bebef4fcde84ac6b.rmeta --extern itertools=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libitertools-e765cbb36c05ba8d.rmeta --extern lazy_static=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/liblazy_static-21bdfe15b12d5d72.rmeta --extern log=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/liblog-ef6f78e2c8f11cdd.rmeta --extern macro_rules_attribute=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libmacro_rules_attribute-d422d9ba04ea3455.rmeta --extern monostate=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libmonostate-098b7fc2d6ce78b2.rmeta --extern onig=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libonig-6f267bdd630017ff.rmeta --extern paste=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libpaste-f3c191b853ab2697.dylib --extern rand=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/librand-ac52eb958f7428d9.rmeta --extern rayon=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/librayon-6c20fc6b04a367e3.rmeta --extern rayon_cond=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/librayon_cond-880e9ecfaec989d3.rmeta --extern regex=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libregex-3717e36dd46ae176.rmeta --extern regex_syntax=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libregex_syntax-a2b1e3a1b590488e.rmeta --extern reqwest=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libreqwest-5dd1ccff69187c61.rmeta --extern serde=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libserde-e04845fb1bbf17c2.rmeta --extern serde_json=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libserde_json-a34e0ef09e0589ad.rmeta --extern spm_precompiled=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libspm_precompiled-abe5d9a77845f07f.rmeta --extern thiserror=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libthiserror-af84165eb92699fc.rmeta --extern unicode_normalization_alignments=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libunicode_normalization_alignments-8813c08cb17332a7.rmeta --extern unicode_segmentation=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libunicode_segmentation-b16e74da88d7993c.rmeta --extern unicode_categories=/Users/piyushpande/tokenizers/bindings/python/target/release/deps/libunicode_categories-e862cc8f2d923063.rmeta -C link-arg=-undefined -C link-arg=dynamic_lookup -C link-arg=-mmacosx-version-min=10.11 -L native=/Users/piyushpande/tokenizers/bindings/python/target/release/build/bzip2-sys-54a51327695825c4/out/lib -L native=/Users/piyushpande/tokenizers/bindings/python/target/release/build/zstd-sys-6efba377626a7cc7/out -L native=/Users/piyushpande/tokenizers/bindings/python/target/release/build/esaxx-rs-d5c73a9453fa2f5a/out -L native=/Users/piyushpande/tokenizers/bindings/python/target/release/build/onig_sys-4fb47a3e4cbb91a0/out` (exit status: 1)
error: `cargo rustc --lib --message-format=json-render-diagnostics --manifest-path Cargo.toml --release -v --features pyo3/extension-module --crate-type cdylib -- -C 'link-args=-undefined dynamic_lookup -Wl,-install_name,@rpath/tokenizers.cpython-39-darwin.so'` failed with code 101
(MultiModular_transformer) piyushpande@Piyushs-MacBook-Pro python %
